sw/device/lib/handler.h

```diff
@@ -16,4 +16,59 @@ typedef enum exc_id {
   kIdMax = 31
 } exc_id_t;
 
+/**
+ * Default exception handler. Can be overidden.
+ */
+__attribute__((aligned(4), interrupt)) void handler_exception(void);
+
+/**
+ * SW IRQ handler. Can be overidden.
+ */
+__attribute__((aligned(4), interrupt)) void handler_irq_software(void);
+
+/**
+ * Timer IRQ handler. Can be overidden.
+ */
+__attribute__((aligned(4), interrupt)) void handler_irq_timer(void);
+
+/**
+ * external IRQ handler. Can be overidden.
+ */
+__attribute__((aligned(4), interrupt)) void handler_irq_external(void);
+
+/**
+ * Instruction access fault. Can be overriden.
+ *
+ * Called by default implementation of handler_exception.
+ */
+void handler_instr_acc_fault(void);
+
+/**
+ * Illegal Instruction fault. Can be overriden.
+ *
+ * Called by default implementation of handler_exception.
+ */
+void handler_instr_ill_fault(void);
+
+/**
+ * Breakpoint handler. Can be overriden.
+ *
+ * Called by default implementation of handler_exception.
+ */
+void handler_bkpt(void);
+
+/**
+ * Load store unit fault. Can be overriden.
+ *
+ * Called by default implementation of handler_exception.
+ */
+void handler_lsu_fault(void);
+
+/**
+ * Exception call handler. Can be overriden.
+ *
+ * Called by default implementation of handler_exception.
+ */
+void handler_ecall(void);
```

Comment:
How come these aren't `__attribute__((interrupt))`?  I know they weren't in the source material, but now might be the time to fix them.

Reply:
`__attribute__((interrupt))` merely changes the epilogs and prologs, so that a whole load of extra registers are saved when an interrupt happens (you need to save almost all the registers, not just the callee-saved registers). This is only needed on the first function that is invoked by the core when an interrupt happens.

In particular, these functions that now lack the attribute are not called by the core when an interrupt happens, they are called by `handler_exception` in `handle.c`, so they don't need the special calling convention.

`handler_exception` does need the special calling convention, and it still has `__attribute__((interrupt))`.

Reply:
Ah, I was missing that they were called by `handler_exception` and not interrupt handlers themselves.  Thanks for the explanation.

Reply:
I question whether doing weak symbols for exception handlers is the right solution; it makes sense for interrupts, due to Ibex's vectoring mechanism, but I think that the exception handlers are best held in a global vtable somewhere (like in that other PR I did).

Reply:
I think there are pros and cons to using either weak symbols or using a vtable. We can discuss it on your other PR.